File: histogram.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
using namespace std;

// FILE OPENING
ifstream fin;
ofstream fout;
string InputFileName;
string OutputFileName;
void setIOOpen(string name) {

	InputFileName = name + ".txt";
	OutputFileName = "cpp_histogram.txt";

	fout.open(OutputFileName, ios::out | ios::trunc);

	if (!fout) {
		fout << "Unable to open file \n";
		exit(1); // terminate with error
	}

	fin.open(InputFileName, ios::in);

	if (!fin) {
		fout << "Unable to open file \n";
		exit(1); // terminate with error
	}
}
//FILE CLOSING
void setIOClose()
{ 
	fin.close();
	fout.close();
}


int numInterval;
int amountData;

vector<pair<int,int>> intervals;
vector<int> dataHolder;

void read()
{
    int a; //reads in the number of intervals
	fin >> a;
	numInterval = a;

	int b; //reads in the number of data values
	fin >> b;
	amountData = b;
    
    for(int i = 0; i < numInterval; i++) { // reads in the intervals in form of pairs
		int start;
		int end;
		fin >> start >> end;
		intervals.push_back( make_pair(start, end) );
	}

	for(int i = 0; i < amountData; i++) { //reads in the raw data itself
		int value;
		fin >> value;
		dataHolder.push_back(value);
	}

	sort(dataHolder.begin(), dataHolder.end()); //sort the data
}

void build() 
{
	cout << "Test";
	vector<int> intervalFreqs;

    for (int i = 0; i < intervals.size(); i++) { //goes through the intervals
		int start = intervals[i].first; // start of interval
		int end = intervals[i].second; //end of interval

		int frequency = 0; //how many data values are in current interval

		for (int value: dataHolder) {
			if(start <= value && value <= end) {
				frequency++;
			}
		}

		intervalFreqs.push_back(frequency); // enters frequency for the interval
	}

	// MAKING HISTOGRAM
	cout << "DATA HISTOGRAM";
    for (int i = 0; i < intervals.size(); i++) { //goes through the intervals
		int start = intervals[i].first; // start of interval
		int end = intervals[i].second; //end of interval
		fout << start << "-" << end;

		fout.width(2); 
        fout << " | "; 
		
		int freq = intervalFreqs[i];
		for(int j = 0; j < freq; j++) {
			fout << " x "; 
		}

		fout << endl;

	}
    fout << "---------------------------------------" << endl; 
    fout << "    "; 
  
  
}

int main()
{
	setIOOpen("data");
	read();
    build();
    setIOClose();
    return 0;
}

File: histogram.py
import os

# FILE OPENING
def fileOpen(name):
    input_file = name + ".txt"

    try:
        data_input = open(input_file, 'rt') #read in data in text form
    except FileNotFoundError:
        print("Error, Input")
        exit(1) 

    try:
        data_output = open("py_histogram.txt", 'w') #write out to here
    except FileNotFoundError:
        print("Error, Output")
        exit(1) 
    
    return (data_input, data_output)

# FILE CLOSING
def fileClose(data_input, data_output):
    data_input.close()
    data_output.close()

intervals = []
data_holder = []
    
def read(data_input):

    line_one = data_input.readline().strip() # reads in the number of intervals
    num_interval = int(line_one)  
    line_two = data_input.readline().strip() # reads in the amount of data
    amount_data = int(line_two) 
    
    for _ in range(num_interval):  
        next_line = data_input.readline().strip().split() #read in the pair
        start, end = map(int, next_line)
        intervals.append((start, end))



    data_line = data_input.readline().strip()  # reads the raw data line
    data_values = data_line.split()  # split the data line into individual values

    for value in data_values:
        data_holder.append((int)(value))

    data_holder.sort()  # sort the data

def build(data_output):
    interval_freqs = []

    for start, end in intervals:  # goes through the intervals
        frequency = sum(start <= value <= end for value in data_holder)  # how many data values are in current interval
        interval_freqs.append(frequency)  # enters frequency for the interval

    # MAKING HISTOGRAM
    data_output.write("DATA HISTOGRAM \n")
    for (start, end), freq in zip(intervals, interval_freqs):  # goes through the intervals
        data_output.write(f"{start}-{end} | {' x ' * freq}\n")
    
    data_output.write("---------------------------------------\n")


def main():
    data_input, data_output = fileOpen("data2")
    read(data_input)
    build(data_output)
    fileClose(data_input, data_output)

if __name__ == "__main__":
    main()
